src/poll.rs

```diff
@@ -564,107 +967,122 @@ impl Drop for RegistrationInner {
-                // barrier before reading events, which will be read using
-                // `Relaxed` ordering. Reading events w/ `Relaxed` is OK thanks to
-                // the acquire / release hand off on `queued`.
-                let mut queued = node_ref.queued.load(Ordering::Acquire);
-                events = node_ref.poll_events();
-
-                // Enter a loop attempting to unset the "queued" bit or requeuing
-                // the node.
-                loop {
-                    // In the following conditions, the registration is removed from
-                    // the readiness queue:
-                    //
-                    // - The registration is edge triggered.
-                    // - The event set contains no events
-                    // - There is a requested delay that has not already expired.
-                    //
-                    // If the drop flag is set though, the node is never queued
-                    // again.
-                    if event::is_drop(events) {
-                        // dropped nodes are always processed immediately. There is
-                        // also no need to unset the queued bit as the node should
-                        // not change anymore.
-                        break;
-                    } else if opts.is_edge() || event::is_empty(events) {
-                        // An acquire barrier is set in order to re-read the
-                        // `events field. `Release` is not needed as we have not
-                        // mutated any field that we need to expose to the producer
-                        // thread.
-                        let next = node_ref.queued.compare_and_swap(queued, 0, Ordering::Acquire);
-
-                        // Re-read in order to ensure we have the latest value
-                        // after having marked the registration has dequeued from
-                        // the readiness queue. Again, `Relaxed` is OK since we set
-                        // the barrier above.
-                        events = node_ref.poll_events();
-
-                        if queued == next {
-                            break;
-                        }
+        // `until` is set with the first node that gets re-enqueued due to being
+        // set to have level-triggered notifications. This prevents an infinite
+        // loop where `Poll::poll` will keep dequeuing nodes it enqueues.
+        let mut until = ptr::null_mut();
+
+        'outer:
+        while dst.len() < dst.capacity() {
+            // Dequeue a node. If the queue is in an inconsistent state, then
+            // stop polling. `Poll::poll` will be called again shortly and enter
+            // a syscall, which should be enough to enable the other thread to
+            // finish the queuing process.
+            let ptr = match self.dequeue_node(until) {
+                Dequeue::Empty | Dequeue::Inconsistent => break,
```

Comment:
This handling of `Inconsistent` will make for some interesting documentation I think. We basically need to document zero guarantees about the return value of poll. This clause means that code like this may fail:

```rust
node.set_readiness(Read::readable());
poll.poll(&mut events, Some(Duration::new(0))); // may return 0
```

That is, this inconsistent state could be induced by another thread which blocks all other threads from having their notifications seen, including the current thread.

Now those semantics seem totally ok to me as any practical usage of mio is just spinning in `poll` anyway, but we may want to be careful (when documenting everything) to mention this.